mm2src/kdf_walletconnect/src/connection_handler.rs

```diff
@@ -0,0 +1,137 @@
+
+const INITIAL_RETRY_SECS: f64 = 5.0;
+const MAX_BACKOFF: u64 = 60;
+const RETRY_INCREMENT: f64 = 5.0;
+
+pub struct Handler {
+    name: &'static str,
+    msg_sender: UnboundedSender<PublishedMessage>,
+    conn_live_sender: UnboundedSender<Option<String>>,
+}
+
+impl Handler {
+    pub fn new(
+        name: &'static str,
+        msg_sender: UnboundedSender<PublishedMessage>,
+        conn_live_sender: UnboundedSender<Option<String>>,
+    ) -> Self {
+        Self {
+            name,
+            msg_sender,
+            conn_live_sender,
+        }
+    }
+}
+
+impl ConnectionHandler for Handler {
+    fn connected(&mut self) {
+        debug!("[{}] connection to WalletConnect relay server successful", self.name);
+    }
+
+    fn disconnected(&mut self, frame: Option<CloseFrame<'static>>) {
+        debug!("[{}] connection closed: frame={frame:?}", self.name);
+
+        if let Err(e) = self.conn_live_sender.unbounded_send(frame.map(|f| f.to_string())) {
+            error!("[{}] failed to send to the receiver: {e}", self.name);
+        }
+    }
+
+    fn message_received(&mut self, message: PublishedMessage) {
+        debug!(
+            "[{}] inbound message: message_id={} topic={} tag={} message={}",
+            self.name, message.message_id, message.topic, message.tag, message.message,
+        );
+
+        if let Err(e) = self.msg_sender.unbounded_send(message) {
+            error!("[{}] failed to send to the receiver: {e}", self.name);
+        }
+    }
+
+    fn inbound_error(&mut self, error: ClientError) {
+        debug!("[{}] inbound error: {error}", self.name);
+        if let Err(e) = self.conn_live_sender.unbounded_send(Some(error.to_string())) {
+            error!("[{}] failed to send to the receiver: {e}", self.name);
+        }
+    }
+
+    fn outbound_error(&mut self, error: ClientError) {
+        debug!("[{}] outbound error: {error}", self.name);
+        if let Err(e) = self.conn_live_sender.unbounded_send(Some(error.to_string())) {
+            error!("[{}] failed to send to the receiver: {e}", self.name);
+        }
+    }
+}
+
+/// Establishes initial connection to WalletConnect relay server with linear retry mechanism.
+/// Uses increasing delay between retry attempts starting from INITIAL_RETRY_SECS.
+/// After successful connection, attempts to restore previous session state from storage.
+pub(crate) async fn spawn_connection_initialization(
+    wc: Arc<WalletConnectCtxImpl>,
+    connection_live_rx: UnboundedReceiver<Option<String>>,
+) {
+    info!("Initializing WalletConnect connection");
+    let mut retry_count = 0;
+    let mut retry_secs = INITIAL_RETRY_SECS;
+
+    while let Err(err) = wc.connect_client().await {
+        retry_count += 1;
+        error!(
+            "Error during initial connection attempt {}: {:?}. Retrying in {retry_secs} seconds...",
+            retry_count, err
+        );
+        Timer::sleep(retry_secs).await;
+        retry_secs += RETRY_INCREMENT;
```

Comment:
i think we should cap this, just like the backoff mechanism below. (it's increasing linearly, so less chance to shoot up, but better safe than sorry).

Reply:
yep, updated to increase exponentiallly